### PR TITLE
Implementation of AddGridTransform

### PR DESCRIPTION
**Pull request by liopeer**
## Motivation
`AddGridTransform` is meant to segment the image into local patches based on a regular grid and works on arbitrary data structures that contain `BoundingBoxes` and `Mask` instances (see `torchvision.tv_tensors` and `torchvision.transforms.v2`) following DetCon ([https://arxiv.org/abs/2103.10957](https://arxiv.org/abs/2103.10957)).

## Changes
1. Implements the transform such that
  - overwrites `Masks`s to create regular grid of classes
  - overwrites `BoundingBoxes`' to create regular grid of bounding boxes
  - leaves anything else in the data structure untouched
2. tests transform specifically against
  - image shape non-divisible by number of columns/rows
3. makes transform public in `transforms/__init__.py`
4. adds the transform to the docs (screenshot below)

## Demo
below a plot for an image size of (32, 32), mask is on the left, drawn bounding boxes on the right
addgrid

## Docs
Screenshot 2024-11-12 at 17 10 40

### CODE REVIEW

File: lightly/transforms/add_grid_transform.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
+#
```
Comment (philippmwirth) [inline lightly/transforms/add_grid_transform.py:1]: I think we use a different license header for `lightly`, see e.g. [here](https://github.com/lightly-ai/lightly/blob/master/lightly/models/utils.py).

**Review by philippmwirth — CHANGES_REQUESTED** [review]

File: lightly/transforms/add_grid_transform.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+class AddGridTransform(Transform):  # type: ignore[misc]
+    """Implements the naive segmentation into a regular grid from DetCon. [0]
+
+    Input to this transform:
+        Any datastructure containing one or several `torchvision.tv_tensor.BoundingBoxes`
+        and/or `torchvision.tv_tensor.Mask`. Masks should be of size (*, H, W) and
+        BoundingBoxes can be of arbitrary shape.
+
+    Output of this transform:
+        Leaves any images in the data structure untouched, but overwrites any bounding
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py:28]: Correct, the user has to provide empty/randomly initialized bounding boxes and masks inside the data structure that will be overwritten (maybe this should be made clearer). The v2 transforms should (if possible) work on arbitrary data structures, and return that exact data structure, therefore altering existing bounding boxes/masks is easier than creating them – since we don't know where we should insert them. Also see:

[https://pytorch.org/vision/main/auto_examples/transforms/plot_transforms_getting_started.html#what-do-i-pass-as-input](https://pytorch.org/vision/main/auto_examples/transforms/plot_transforms_getting_started.html#what-do-i-pass-as-input)

File: lightly/transforms/add_grid_transform.py (hunk 1 above)
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py]: fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/c91fdb81474749bff9c499877062643f19bc40b1

File: lightly/transforms/add_grid_transform.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+from torchvision.tv_tensors import BoundingBoxes, BoundingBoxFormat, Mask
+
+from lightly.utils import dependency as _dependency
+
+
+# ignore typing due to Any type used in torchvison.transforms.v2.Transform
+class AddGridTransform(Transform):  # type: ignore[misc]
+    """Implements the naive segmentation into a regular grid from DetCon. [0]
+
+    Input to this transform:
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py:17]: I was mainly following the other transforms, e.g. [BYOL](https://github.com/lightly-ai/lightly/blob/master/lightly/transforms/byol_transform.py), [MAE](https://github.com/lightly-ai/lightly/blob/master/lightly/transforms/mae_transform.py). But I am more than happy to start transitioning to `Args:` and `Returns:`, maybe even do it a bit more proper and put them inside a separate docstring of the `__call__` or `forward` method.

File: lightly/transforms/add_grid_transform.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(self, num_rows: int, num_cols: int) -> None:
+        if not _dependency.torchvision_transforms_v2_available():
+            raise ImportError(
+                "AddGridTransform requires torchvision.transforms.v2, included in torchvision>=0.17"
+            )
+        super().__init__()
+        self.num_rows = num_rows
+        self.num_cols = num_cols
+
+    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py:56]: Yes, required by the parent class [torchvision.transforms.v2.Transform](https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/_transform.py).

File: lightly/transforms/add_grid_transform.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Any datastructure containing one or several `torchvision.tv_tensor.BoundingBoxes`
+        and/or `torchvision.tv_tensor.Mask`. Masks should be of size (*, H, W) and
+        BoundingBoxes can be of arbitrary shape.
+
+    Output of this transform:
+        Leaves any images in the data structure untouched, but overwrites any bounding
+        boxes and masks by a regular grid. Bounding boxes will take shape (num_rows*num_cols, 4)
+        and masks will be of shape (1, H, W) with integer values in the range [0, num_rows*num_cols-1].
+
+    Useage example:
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py]: fixed

File: lightly/transforms/add_grid_transform.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+from lightly.utils import dependency as _dependency
+
+
+# ignore typing due to Any type used in torchvison.transforms.v2.Transform
+class AddGridTransform(Transform):  # type: ignore[misc]
+    """Implements the naive segmentation into a regular grid from DetCon. [0]
+
+    Input to this transform:
+        Any datastructure containing one or several `torchvision.tv_tensor.BoundingBoxes`
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py:18]: fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/af3a02b2c21a6073c451716ccf5801d46fde172c

File: lightly/transforms/add_grid_transform.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                "AddGridTransform requires torchvision.transforms.v2, included in torchvision>=0.17"
+            )
+        super().__init__()
+        self.num_rows = num_rows
+        self.num_cols = num_cols
+
+    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
+        if isinstance(inpt, BoundingBoxes):
+            return _create_bounding_boxes_grid(
+                self.num_rows,
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py]: fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/038ec59e48112417c07310752543149dd8b28881

File: lightly/transforms/add_grid_transform.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                dtype=inpt.dtype,
+                device=inpt.device,
+                requires_grad=inpt.requires_grad,
+                format=inpt.format,
+            )
+        elif isinstance(inpt, Mask):
+            return _create_mask_grid(
+                self.num_rows,
+                self.num_cols,
+                canvas_size=inpt.shape[-2:],
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/transforms/add_grid_transform.py:76]: fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/d64264fa4cefc0e497aec423672c7b32d760a53f

File: tests/transforms/test_add_grid_transform.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            [3.0, 3.0, 4.0, 4.0, 5.0],
+            [3.0, 3.0, 4.0, 4.0, 5.0],
+            [3.0, 3.0, 4.0, 4.0, 5.0],
+        ]
+    ).to(torch.int64)
+    return Mask(mask)
+
+
+@pytest.fixture
+def cols() -> int:
```
[the remark this answers is not in the document]
Comment (liopeer) [reply tests/transforms/test_add_grid_transform.py]: I agree. Initially thought that I would use them in all tests. Fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/2cd76cb42b92d4231b28b8c97bfe3bbb98841876

File: tests/transforms/test_add_grid_transform.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+# ignore typing due to Any type used in torchvison.transforms.v2.Transform
+@pytest.fixture
+def img_orig() -> PILImage:  # type: ignore[misc]
+    img = PILImageModule.new("RGB", (7, 5))
+    return img
+
+
+# ignore typing due to Any type used in torchvison.transforms.v2.Transform
+@pytest.fixture
+def bbox_expected() -> BoundingBoxes:  # type: ignore[misc]
```
[the remark this answers is not in the document]
Comment (liopeer) [reply tests/transforms/test_add_grid_transform.py]: fixed in https://github.com/lightly-ai/lightly/commit/2cd76cb42b92d4231b28b8c97bfe3bbb98841876

File: lightly/transforms/add_grid_transform.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (philippmwirth) [reply lightly/transforms/add_grid_transform.py:17]: I see, note that these transforms are special because they combine multiple transforms into one for convenience. The transform you're implementing here however is standalone (see e.g. `rotation.py` for reference). However, you're right that we don't have a `__call__` or `forward` function to annotate so let's keep it the way it is for now.

File: tests/transforms/test_add_grid_transform.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+    tr = AddGridTransform(rows, cols)
+    bbox_empty = BoundingBoxes(
+        torch.zeros(1, 4), format="XYXY", canvas_size=(img_h, img_w)
+    )
+    bbox_tr = tr(bbox_empty)
+    assert_close(bbox_expected, bbox_tr, atol=0.01, rtol=0.01)
+
+
+def test_AddGridTransform_mask(cols: int, rows: int) -> None:
```
Comment (philippmwirth) [inline tests/transforms/test_add_grid_transform.py:56]: Very minor: The order of `cols` and `rows` argument is inconsistent between tests.

**Review by philippmwirth — APPROVED** [review]

File: tests/transforms/test_add_grid_transform.py (hunk 11 above)
Comment (liopeer) [reply tests/transforms/test_add_grid_transform.py:56]: fixed in https://github.com/lightly-ai/lightly/pull/1731/commits/ccaa553d6e42d5ac1d6fdce29eac068e37f9f892
